File: 3-HARD/Python/4-segment-tree-lazy-propagation/proposed-solution/segment_tree.py
from typing import List

class SegmentTree:
    def __init__(self, data: List[int]):
        self.n = len(data)
        self.tree = [0] * (4 * self.n)
        self.lazy = [0] * (4 * self.n)
        self._build(data, 0, 0, self.n - 1)

    def _build(self, data: List[int], node: int, start: int, end: int):
        if start == end:
            self.tree[node] = data[start]
        else:
            mid = (start + end) // 2
            self._build(data, 2 * node + 1, start, mid)
            self._build(data, 2 * node + 2, mid + 1, end)
            self.tree[node] = self.tree[2 * node + 1] + self.tree[2 * node + 2]

    def _update_range(self, node: int, start: int, end: int, l: int, r: int, val: int):
        if self.lazy[node] != 0:
            self.tree[node] += (end - start + 1) * self.lazy[node]
            if start != end:
                self.lazy[2 * node + 1] += self.lazy[node]
                self.lazy[2 * node + 2] += self.lazy[node]
            self.lazy[node] = 0

        if start > r or end < l:
            return

        if l <= start and end <= r:
            self.tree[node] += (end - start + 1) * val
            if start != end:
                self.lazy[2 * node + 1] += val
                self.lazy[2 * node + 2] += val
            return

        mid = (start + end) // 2
        self._update_range(2 * node + 1, start, mid, l, r, val)
        self._update_range(2 * node + 2, mid + 1, end, l, r, val)
        self.tree[node] = self.tree[2 * node + 1] + self.tree[2 * node + 2]

    def update_range(self, l: int, r: int, val: int):
        self._update_range(0, 0, self.n - 1, l, r, val)

    def _query_range(self, node: int, start: int, end: int, l: int, r: int) -> int:
        if start > r or end < l:
            return 0

        if self.lazy[node] != 0:
            self.tree[node] += (end - start + 1) * self.lazy[node]
            if start != end:
                self.lazy[2 * node + 1] += self.lazy[node]
                self.lazy[2 * node + 2] += self.lazy[node]
            self.lazy[node] = 0

        if l <= start and end <= r:
            return self.tree[node]

        mid = (start + end) // 2
        left_sum = self._query_range(2 * node + 1, start, mid, l, r)
        right_sum = self._query_range(2 * node + 2, mid + 1, end, l, r)
        return left_sum + right_sum

    def query_range(self, l: int, r: int) -> int:
        return self._query_range(0, 0, self.n - 1, l, r)

File: 3-HARD/JavaScript/4-segment-tree-lazy-propagation/proposed-solution/SegmentTree.js
// SegmentTree.js - Segment Tree with Lazy Propagation in JavaScript

class SegmentTree {
  constructor(data) {
    this.n = data.length;
    this.tree = new Array(4 * this.n).fill(0);
    this.lazy = new Array(4 * this.n).fill(0);
    this._build(data, 0, 0, this.n - 1);
  }

  _build(data, node, start, end) {
    if (start === end) {
      this.tree[node] = data[start];
    } else {
      const mid = Math.floor((start + end) / 2);
      this._build(data, 2 * node + 1, start, mid);
      this._build(data, 2 * node + 2, mid + 1, end);
      this.tree[node] = this.tree[2 * node + 1] + this.tree[2 * node + 2];
    }
  }

  _updateRange(node, start, end, l, r, val) {
    if (this.lazy[node] !== 0) {
      this.tree[node] += (end - start + 1) * this.lazy[node];
      if (start !== end) {
        this.lazy[2 * node + 1] += this.lazy[node];
        this.lazy[2 * node + 2] += this.lazy[node];
      }
      this.lazy[node] = 0;
    }

    if (start > r || end < l) {
      return;
    }

    if (l <= start && end <= r) {
      this.tree[node] += (end - start + 1) * val;
      if (start !== end) {
        this.lazy[2 * node + 1] += val;
        this.lazy[2 * node + 2] += val;
      }
      return;
    }

    const mid = Math.floor((start + end) / 2);
    this._updateRange(2 * node + 1, start, mid, l, r, val);
    this._updateRange(2 * node + 2, mid + 1, end, l, r, val);
    this.tree[node] = this.tree[2 * node + 1] + this.tree[2 * node + 2];
  }

  updateRange(l, r, val) {
    this._updateRange(0, 0, this.n - 1, l, r, val);
  }

  _queryRange(node, start, end, l, r) {
    if (start > r || end < l) {
      return 0;
    }

    if (this.lazy[node] !== 0) {
      this.tree[node] += (end - start + 1) * this.lazy[node];
      if (start !== end) {
        this.lazy[2 * node + 1] += this.lazy[node];
        this.lazy[2 * node + 2] += this.lazy[node];
      }
      this.lazy[node] = 0;
    }

    if (l <= start && end <= r) {
      return this.tree[node];
    }

    const mid = Math.floor((start + end) / 2);
    const leftSum = this._queryRange(2 * node + 1, start, mid, l, r);
    const rightSum = this._queryRange(2 * node + 2, mid + 1, end, l, r);
    return leftSum + rightSum;
  }

  queryRange(l, r) {
    return this._queryRange(0, 0, this.n - 1, l, r);
  }
}

export default SegmentTree;
